message_ix/utils.py

```diff
@@ -16,6 +21,105 @@ def logger():
     return _LOGGER
 
 
+def isstr(x):
+    """Returns True if x is a string"""
+    return isinstance(x, six.string_types)
+
+
+def isscalar(x):
+    """Returns True if x is a scalar"""
+    return not isinstance(x, collections.Iterable) or isstr(x)
+
+
+def add_spatial_sets(scenario, data):
+    """Add sets related to spatial dimensions of the model
+
+    Parameters
+    ----------
+    scenario : ixmp.Scenario
+    data : dict or other
+
+    Examples
+    --------
+    data = {'country': 'Austria'}
+    data = {'country': ['Austria', 'Germany']}
+    data = {'country': {'Austria': {'state': ['Vienna', 'Lower Austria']}}}
+    """
+    nodes = []
+    levels = []
+    hierarchy = []
+
+    def recurse(k, v, parent='World'):
+        if isinstance(v, collections.Mapping):
+            for _parent, _data in v.items():
+                for _k, _v in _data.items():
+                    recurse(_k, _v, parent=_parent)
+
+        level = k
+        children = [v] if isscalar(v) else v
+        for child in children:
+            hierarchy.append([level, child, parent])
+            nodes.append(child)
+        levels.append(level)
+
+    for k, v in data.items():
+        recurse(k, v)
+
+    scenario.add_set("node", nodes)
+    scenario.add_set("lvl_spatial", levels)
+    scenario.add_set("map_spatial_hierarchy", hierarchy)
+
+
+def add_temporal_sets(scenario, data):
+    """Add sets related to temporal dimensions of the model
+
+    Parameters
+    ----------
+    scenario : ixmp.Scenario
+    data : dict or other
+
+    Examples
+    --------
+    data = {'year': [2010, 2020]}
+    data = {'year': [2010, 2020], 'firstmodelyear': 2020}
+    """
+    if 'year' not in data:
+        raise ValueError('"year" must be in temporal sets')
+    horizon = data['year']
+    scenario.add_set("year", horizon)
+
+    first = data['firstmodelyear'] if 'firstmodelyear' in data else horizon[0]
+    scenario.add_set("cat_year", ["firstmodelyear", first])
+
+
+def vintage_and_active_years(scenario):
```

Comment:
This approach creates (in most cases) far more data points than required - verbose and not not elegant. As an alternative, I'd suggest to use the function `years_active(node, tec, vintage)`, which returns the list of years in which this `technology` at a particular `node` built in year `vintage` can be active. On the contra side, this requires the parameter `technical_lifetime` to be set first - on the plus side, it takes into account changing period duration.

Reply:
We use this all throughout the tutorial. My goal here was simply reducing complexity there. I believe you yourself said this was distracting when message was taught in italy, right?

Reply:
When you implemented the first version of the Austria tutorial, the `years_active()` feature wasn't implemented yet.